src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,340 @@
+        Stmt body = mutate(op->body);
+        Expr extent = Expr(1);
+        for (Expr e : op->extents) {
+            extent = extent * e;
+        }
+        body = allocate_mutex(mutex_name, extent, body);
+        return Allocate::make(op->name,
+                              op->type,
+                              op->memory_type,
+                              op->extents,
+                              op->condition,
+                              std::move(body),
+                              op->new_expr,
+                              op->free_function);
+    }
+
+    Stmt visit(const ProducerConsumer *op) override {
+        // Usually we allocate the mutex buffer at the Allocate node,
+        // but outputs don't have Allocate. For those we allocate the mutex
+        // buffer at the producer node.
+
+        if (!op->is_producer) {
+            // This is a consumer.
+            return IRMutator::visit(op);
+        }
+
+        // Find the corresponding output.
+        auto func_it = env.find(op->name);
+        Func f = Func(func_it->second);
+        internal_assert(f.output_buffers().size() > 0) <<
+            "Found a producer node that contains an atomic node that requires mutex lock, "
+            "but does not have an Allocate node and is not an output function. This is not supported.\n";
+
+        set<string> store_names;
+        for (auto buffer : f.output_buffers()) {
+            store_names.insert(buffer.name());
+        }
+
+        FindStoreInAtomicMutex finder(store_names);
+        op->body.accept(&finder);
+        if (!finder.found) {
+            // No Atomic node that requires mutex lock from this node inside.
+            return IRMutator::visit(op);
+        }
+
+        if (allocated_mutexes.find(finder.mutex_name) != allocated_mutexes.end()) {
+            // We've already allocated a mutex.
+            return IRMutator::visit(op);
+        }
+
+        allocated_mutexes.insert(finder.mutex_name);
+
+        // We assume all output buffers in a Tuple have the same extent.
+        OutputImageParam output_buffer = f.output_buffers()[0];
+        Expr extent = Expr(1);
+        for (int i = 0; i < output_buffer.dimensions(); i++) {
+            extent = extent * output_buffer.dim(i).extent();
+        }
+        Stmt body = mutate(op->body);
+        body = allocate_mutex(finder.mutex_name, extent, body);
+        return ProducerConsumer::make(op->name, op->is_producer, std::move(body));
+    }
+
+    Stmt visit(const Atomic *op) override {
+        if (op->mutex_name.empty()) {
+            return IRMutator::visit(op);
+        }
+
+        // Lock the mutexes using the indices from the Store nodes inside the body.
+        FindStoreIndex find;
+        op->body.accept(&find);
+
+        Expr index = find.index;
+        if (!index.defined()) {
+            // scalar output
+            index = Expr(0);
+        }
+        Stmt body = op->body;
```

Comment:
Consider:
```

    f(x) = 0;
    f(clamp(f(r), 0, 100)) = f(r) + 1;

```
You need a check:
```
user_assert(!expr_uses_var(index, name of the buffer)) 
    << "Can't use atomic() on an update where the index written to depends on the current value of the Func";
```

Reply:
Done. I did this in a separate pass (see CheckAtomicValidity) since it also applies to atomic stores. Also added a test in error.